### ISSUE TITLE
`twistd web` should work on Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8008]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#8008|
|Type|enhancement|
|Created|2015-08-30 05:22:19Z|
|Branch|https://github.com/twisted/twisted/tree/twistd-web-py3-8008-3|

Tests pass, and I manually install it and it works -- please review.

<details><summary>Searchable metadata</summary>

```
trac-id__8008 8008
type__enhancement enhancement
reporter__hawkowl hawkowl
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_twistd_web_py3_8008_3 branches/twistd-web-py3-8008-3
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1440912139879552 1440912139879552
changetime__1441595684167687 1441595684167687
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [45569]) Branching to twistd-web-py3-8008.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

Tests pass, but it needs some more tests to bump up coverage...

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

(In [45582]) Branching to twistd-web-py3-8008-2.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Thanks for working on this.

Changes look good.

Maybe this can be split into 2 tickets: One for porting the twistd script and another one for the partial port of the web plugin.

I see that twcgi, wsgi and personal folders are not yet ported.  Maybe the title/description of this ticket should be updated to inform that not all the functionality of twistd web was ported.

The comments containing the trac tickets should be formatted as a FIXME commend according to the coding guide.

There is no newsfile fragment.

Please check my comments and resubmit... I hope that we can move twistd port into a separate ticket.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban commented|
|-|-|

[#8012](https://github.com/twisted/twisted/issues/8012) was created to focus on porting twistd script and let people know about it :)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl removed owner|
|-|-|

I've pulled out the twistd changes, and done the required changes.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl commented|
|-|-|

(In [45607]) Branching to twistd-web-py3-8008-3.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

Merged forward with the [#8012](https://github.com/twisted/twisted/issues/8012) changes.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_10"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Do we really need to use `endpoints._parseServer` ? Even if we are in the test code, I would prefer to not touch private stuff.

Maybe this is a sign that strport should have not been deprecated yet and create the similar functionality on the endpoints.

-----

For the FIXME comment, I would prefer to have each FIXME + URL on a single line. In this way we can later implement some simple parsers to check that no FIXME ticket was left behind for tickets which were already closed.

Other than that changes look good.

I think that we should try hard not to use private member in 3rd party tests (in this case tests which does not explicitly check endpoints functionality)

Please consider removing the usage to the private member and resubmit.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl removed owner|
|-|-|

Hi Adi,

As per the docstring for the old ```strports.parse```, it has been moved into ```endpoints``` as ```_parseServer``` -- so it's reasonable for us to use it in this context. Overall, Twisted's approach to private stuff is that it's private to people that aren't us.

Fixed up the FIXME comments.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban commented|
|-|-|

> Overall, Twisted's approach to private stuff is that it's private to people that aren't us.

Is this documented somewhere?

I was expecting that private stuff should only be used inside the class and the test dedicated to that class. 

----------

I see [#5540](https://github.com/twisted/twisted/issues/5540) where strports was deprecated.

The deprecation of strports.parse instruct users to use 
twisted.internet.endpoints.serverFromString".
Why are we not doing the same?

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_13"></a>|@hawkowl commented|
|-|-|

The difference between private and non-private code, as per the compat policy, is that we have no guarantees for external compatibility. Eg. ```twisted._threads``` is private, but that's because we're not making any external guarantees about it. Things inside Twisted can still use ```twisted._threads```. The only thing Twisted code isn't allowed to use is things from test modules in regular modules (but test modules can import from other test modules).

In this case, it's because we want to make sure that the endpoint it would make is made with the arguments we want it to. By calling ```_parseServer```, we can get the arguments that the endpoint would be made with, rather than having to actually make an endpoint and then inspect its internal state. We just want to make sure that the parsed endpoint description matches what it should.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_14"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Thanks for the info. All good and we have them as reference, in case someone want to dig into the changes :)

Looking forward for the follow up tickets ... wsgi and stuff :)

Ready to merge.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_15"></a>|@hawkowl set status to `closed`|
|-|-|

(In [45649]) Merge twistd-web-py3-8008-3: Port twistd web to Python 3

Author: hawkowl
Reviewer: adiroiban
Fixes: [#8008](https://github.com/twisted/twisted/issues/8008)
